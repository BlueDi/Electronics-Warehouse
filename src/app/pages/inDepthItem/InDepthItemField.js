import React, { Component } from 'react';
import { ReadjustableImage, HTMLEditor } from '@common/components';
import {
  Dropdown,
  Breadcrumb,
  Input,
  Button,
  Message
} from 'semantic-ui-react';

/**
 * Item image maximum dimensions
 */
const IMAGE_WIDTH = 400;
const IMAGE_HEIGHT = 300;

/**
 * Represents an item information field
 * Exemples: description, image, packaging, properties, etc.
 */
export class InDepthItemField extends Component {
  constructor(props) {
    super(props);
  }

  /**
   * Decides what render function should be called based on the following criteria
   * - Edit mode
   * - Item field is a property or not
   */
  getRender() {
    if (this.props.editable) {
      return this.props.fieldName === 'properties'
        ? this.editablePropertiesField()
        : this.editableGeneralField();
    } else {
      return this.props.fieldName === 'properties'
        ? this.nonEditablePropertiesField()
        : this.nonEditableGeneralField();
    }
  }

  /**
   * Renders all item properties, in non-edit mode
   */
  nonEditablePropertiesField() {
    let property_list = this.props.fieldContent.map(property => {
      let invalidProperty = property.value ? !property.value.trim() : true;

      return (
        <li key={property.property_id}>
          {property.name} : {invalidProperty && 'N/D'}{' '}
          {!invalidProperty && `${property.value}`}{' '}
          {property.unit && !invalidProperty && `${property.unit}`}
        </li>
      );
    });

    let existentSpecifications = (
      <React.Fragment>
        <span>Specifications</span>
        <ul>{property_list}</ul>
      </React.Fragment>
    );
    let noSpecifications = <span>There are no specifications</span>;
    return (
      <div>
        {property_list.length > 0 && existentSpecifications}
        {property_list.length === 0 && noSpecifications}
      </div>
    );
  }

  /**
   * Renders every item field that is not a property, in non-edit mode
   */
  nonEditableGeneralField() {
    switch (this.props.fieldName) {
      case 'image': {
        let image_src = `data:image/png;base64,${this.props.fieldContent}`;
        return (
          <ReadjustableImage
            src={image_src}
            maxWidth={IMAGE_WIDTH}
            maxHeight={IMAGE_HEIGHT}
          />
        );
      }

      case 'description': {
        return (
          <h1 className="Title" style={{ textAlign: 'left', marginLeft: '5%' }}>
            {this.props.fieldContent}
          </h1>
        );
      }

      case 'user comments': {
        return (
          <React.Fragment>
            <div>
              <h3>{this.props.fieldName}</h3>
            </div>
            <HTMLEditor
              displayOnly={true}
              height="250"
              width="625"
              value={this.props.fieldContent}
            />
          </React.Fragment>
        );
      }

      case 'Workflow': {
        return (
          <React.Fragment>
            {this.props.fieldName}: <br />
            <HTMLEditor displayOnly={true} value={this.props.fieldContent} />
          </React.Fragment>
        );
      }

      case 'add user comment': {
        if (!this.props.isUserLogged) {
          //only allow comment addition if user is logged in
          return null;
        }

        return (
          <div style={{ marginTop: 20 }}>
            <HTMLEditor
              className={this.props.fieldName.replace(/ /g, '_')}
              canvasType="code"
              height="100"
              width="625"
              onChange={this.props.handleChange[0]}
              value={this.props.fieldContent}
            />
            <div style={{ marginBottom: 10, position: 'relative' }}>
              <Button
                primary
                key={this.props.fieldName}
                onClick={this.props.handleChange[1]}
              >
                Add comment
              </Button>
            </div>
          </div>
        );
      }

      case 'category': {
        let category_tree = this.props.fieldContent.map((category, index) => {
          let active = index === this.props.fieldContent.length - 1;
          return {
            key: category.id,
            content: category.name,
            link: false,
            active: active
          };
        });
        return (
          <div>
            {this.props.fieldName}:{' '}
            <Breadcrumb icon="right angle" sections={category_tree} />
          </div>
        );
      }

      default:
        return `${this.props.fieldName}: ${this.props.fieldContent}`;
    }
  }

  /**
   * Renders all item properties, in edit mode
   */
  editablePropertiesField() {
    let property_list = this.props.fieldContent.map(property => {
      var type = property.isNumber ? 'number' : 'text';
      return (
        <li key={property.property_id}>
          {property.name} :
          <Input
            type={type}
            name={property.name}
            value={property.value}
            onChange={this.props.handleChange}
          />
          {property.unit}
        </li>
      );
    });

    let existentSpecifications = (
      <React.Fragment>
        <span>Specifications</span>
        <ul>{property_list}</ul>
      </React.Fragment>
    );
    let noSpecifications = <span>There are no specifications</span>;
    return (
      <div>
        {property_list.length > 0 && existentSpecifications}
        {property_list.length === 0 && noSpecifications}
      </div>
    );
  }

  /**
   * Renders every item field that is not a property, in edit mode
   */
  editableGeneralField() {
    switch (this.props.fieldName) {
      case 'description': {
        return (
          <h1 className="Title" style={{ textAlign: 'left', marginLeft: '5%' }}>
            {this.props.fieldName}:{' '}
            <Input
              type="text"
              name={this.props.fieldName}
              value={this.props.fieldContent}
              onChange={this.props.handleChange}
            />
          </h1>
        );
      }

      case 'image': {
        let image_src = `data:image/png;base64,${this.props.fieldContent}`;
        return (
          <div>
            <ReadjustableImage
              src={image_src}
              maxWidth={IMAGE_WIDTH}
              maxHeight={IMAGE_HEIGHT}
            />
            <Button style={{ display: 'block', margin: '0 auto' }}>
              <label htmlFor="files" className="Button">
                Upload Image
              </label>
            </Button>
            <Input
              id="files"
              type="file"
              value=""
              onChange={this.props.handleChange}
              style={{ visibility: 'hidden' }}
            />
          </div>
        );
      }

      case 'user comments': {
        return (
          <div style={{ marginTop: 3 }}>
            <div>
              <h3>{this.props.fieldName}</h3>
            </div>
            <HTMLEditor
              className={this.props.fieldName.replace(/ /g, '_')}
              canvasType="code"
              height="450"
              width="625"
              onChange={this.props.handleChange}
              value={this.props.fieldContent}
            />
          </div>
        );
      }

      case 'packaging': {
        let availablePackages = this.props.fieldContent.packagingList;
        let itemPackaging = this.props.fieldContent.itemPackaging.name;

        return (
          <span>
            {this.props.fieldName}:{' '}
            <Dropdown
              fluid
              search
              selection
              selectOnNavigation={false}
              value={itemPackaging}
              options={availablePackages}
              onChange={this.props.handleChange}
            />
          </span>
        );
      }

      case 'breadcrumb': {
        if (this.props.fieldContent.length === 0) {
          return <Message negative header="WARNING: No category selected!" />;
        }
        let breadcrumb_tree = this.props.fieldContent.map((category, index) => {
          let link = index !== this.props.fieldContent.length - 1;
          let active = index === this.props.fieldContent.length - 1;
          let onClick =
            index === this.props.fieldContent.length - 1
              ? null
              : this.props.handleChange[0];
          return {
            key: category.id,
            content: category.name,
            link: link,
            active: active,
            onClick: onClick
          };
        });

        return (
          <div
            style={{
              padding: '3px 3px',
              backgroundSize: 'auto'
            }}
          >
            category: <br />
            <Breadcrumb icon="right angle" sections={breadcrumb_tree} />
            <Button
              circular
              negative
              icon="delete"
              size="mini"
              onClick={this.props.handleChange[1]}
              style={{
                backgroundColor: '#e82b34',
                marginLeft: '5px',
                padding: '5px 5px'
              }}
            />
          </div>
        );
      }

      case 'category': {
        let availableCategories = this.props.fieldContent.dropdown;
        let itemCategory = this.props.fieldContent.itemCategory.name;

        let noCategoryAvailable = `${itemCategory} has no children`;
        let placeholder = itemCategory
          ? `${itemCategory}'s children`
          : 'Select a category';

        return (
          <span>
            <Dropdown
              fluid
              search
              selection
              selectOnNavigation={false}
              noResultsMessage={noCategoryAvailable}
              text={placeholder}
              options={availableCategories}
              onChange={this.props.handleChange}
            />
          </span>
        );
      }

      case 'Workflow': {
        return (
          <div style={{ marginTop: 3 }}>
            {this.props.fieldName}: <br />
            <HTMLEditor
              canvasType="code"
              onChange={this.props.handleChange}
              value={this.props.fieldContent}
            />
          </div>
        );
      }

      default: {
        return (
          <span>
            {this.props.fieldName}:{' '}
            <Input
              type="text"
              name={this.props.fieldName}
              value={this.props.fieldContent}
              onChange={this.props.handleChange}
            />
          </span>
        );
      }
    }
  }

  /**
   * Default React component render function
   * It's render responsibility is sent to "getRender" function
   */
  render() {
    return this.getRender();
  }
}

export default InDepthItemField;
