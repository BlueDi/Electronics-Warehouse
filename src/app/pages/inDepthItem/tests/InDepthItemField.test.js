import React from 'react';
import renderer from 'react-test-renderer';
import InDepthItemField from '@pages/inDepthItem/InDepthItemField';

describe('<InDepthItemField />', () => {
  test('Renders properly', () => {
    const component = renderer.create(<InDepthItemField />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Breadcrumb Field', () => {
    const myMock = jest.fn();
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={[]}
        fieldName="breadcrumb"
        handleChange={[myMock, myMock]}
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const component2 = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={[
          { key: '1', id: '1', name: 'test' },
          { key: '2', id: '2', name: 'test2' }
        ]}
        fieldName="breadcrumb"
        handleChange={[myMock, myMock]}
      />
    );
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Category Field', () => {
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={{
          breadcrumb: [{ key: 1, id: '1', name: 'test' }],
          categoryList: [
            {
              key: 1,
              text: 'Electronic Components',
              value: 'Electronic Components'
            }
          ],
          dropdown: [
            { key: 1, text: 'c1', value: 'c1' },
            { key: 2, text: 'c2', value: 'c2' }
          ],
          itemCategory: { id: '2', name: '2' }
        }}
        fieldName="category"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const component2 = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={{
          itemCategory: { id: '1' },
          dropdown: [
            { key: 1, text: 'c1', value: 'c1' },
            { key: 2, text: 'c2', value: 'c2' }
          ]
        }}
        fieldName="category"
      />
    );
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot();

    const component3 = renderer.create(
      <InDepthItemField
        editable={false}
        fieldContent={[
          {
            key: 1,
            id: '1',
            name: 'test'
          }
        ]}
        fieldName="category"
      />
    );
    tree = component3.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Description Field', () => {
    var value = 1;
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={value}
        fieldName="description"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const component2 = renderer.create(
      <InDepthItemField
        editable={false}
        fieldContent={value}
        fieldName="description"
      />
    );
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Image Field', () => {
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={
          'iVBORw0KGgoAAAANSUhEUgAAAG8AAABYCAYAAAD7uql1AAAgAElEQVR4Xu29e5RV5Znu+5vfN6/r\nVmtVFXWjoChABFERVCKgUUAjthFzMZqYxLjT3aP79D456d49zj5nZJzTo0f/0+fsHumT7O6xY/be\n2umklcSEaFRuIiiIF0AUFQVFQSAlRRVVRdW6zut3/phrLs…'
        }
        fieldName="image"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const component2 = renderer.create(
      <InDepthItemField
        editable={false}
        fieldContent={
          'iVBORw0KGgoAAAANSUhEUgAAAG8AAABYCAYAAAD7uql1AAAgAElEQVR4Xu29e5RV5Znu+5vfN6/r\nVmtVFXWjoChABFERVCKgUUAjthFzMZqYxLjT3aP79D456d49zj5nZJzTo0f/0+fsHumT7O6xY/be\n2umklcSEaFRuIiiIF0AUFQVFQSAlRRVVRdW6zut3/phrLs…'
        }
        fieldName="image"
      />
    );
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Other Field', () => {
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={'other'}
        fieldName="other"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Packaging Field', () => {
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={{
          itemPackaging: { id: '1', name: 'test' },
          packagingList: [{ key: '1', text: 'p1', value: 'p1' }]
        }}
        fieldName="packaging"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Properties Field', () => {
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={[{ key: 1, name: 'test', value: '1' }]}
        fieldName="properties"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const component2 = renderer.create(
      <InDepthItemField
        editable={false}
        fieldContent={[{ key: 1, name: 'test', value: '1', unit: 'w' }]}
        fieldName="properties"
      />
    );
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Properties Invalid Field', () => {
    const component = renderer.create(
      <InDepthItemField
        editable={true}
        fieldContent={[{ key: 1, name: 'test', isNumber: true }]}
        fieldName="properties"
      />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    const component2 = renderer.create(
      <InDepthItemField
        editable={false}
        fieldContent={[{ key: 1, name: 'test' }]}
        fieldName="properties"
      />
    );
    tree = component2.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
