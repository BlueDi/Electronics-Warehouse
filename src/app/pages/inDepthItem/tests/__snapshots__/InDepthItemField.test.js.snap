// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InDepthItemField /> Breadcrumb Field 1`] = `
<div
  className="ui negative message"
>
  <div
    className="content"
  >
    <div
      className="header"
    >
      WARNING: No category selected!
    </div>
  </div>
</div>
`;

exports[`<InDepthItemField /> Breadcrumb Field 2`] = `
<div
  style={
    Object {
      "backgroundSize": "auto",
      "padding": "3px 3px",
    }
  }
>
  category: 
  <br />
  <div
    className="ui breadcrumb"
  >
    <a
      className="section"
      onClick={[Function]}
    >
      test
    </a>
    <i
      aria-hidden="true"
      className="right angle icon divider"
    />
    <div
      className="active section"
      onClick={[Function]}
    >
      test2
    </div>
  </div>
  <button
    className="ui mini circular icon negative button"
    onClick={[Function]}
    role="button"
    style={
      Object {
        "backgroundColor": "#e82b34",
        "marginLeft": "5px",
        "padding": "5px 5px",
      }
    }
  >
    <i
      aria-hidden="true"
      className="delete icon"
    />
  </button>
</div>
`;

exports[`<InDepthItemField /> Category Field 1`] = `
<span>
  <div
    aria-expanded={false}
    className="ui fluid search selection dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autoComplete="off"
      className="search"
      onChange={[Function]}
      style={
        Object {
          "width": undefined,
        }
      }
      tabIndex={0}
      type="text"
      value=""
    />
    <div
      aria-live="polite"
      className="text"
      role="alert"
    >
      2's children
    </div>
    <i
      aria-hidden="true"
      className="dropdown icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
      role="listbox"
    >
      <div
        aria-checked={false}
        aria-selected={true}
        className="selected item"
        onClick={[Function]}
        role="option"
        style={
          Object {
            "pointerEvents": "all",
          }
        }
      >
        <span
          className="text"
        >
          c1
        </span>
      </div>
      <div
        aria-checked={false}
        aria-selected={false}
        className="item"
        onClick={[Function]}
        role="option"
        style={
          Object {
            "pointerEvents": "all",
          }
        }
      >
        <span
          className="text"
        >
          c2
        </span>
      </div>
    </div>
  </div>
</span>
`;

exports[`<InDepthItemField /> Category Field 2`] = `
<span>
  <div
    aria-expanded={false}
    className="ui fluid search selection dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autoComplete="off"
      className="search"
      onChange={[Function]}
      style={
        Object {
          "width": undefined,
        }
      }
      tabIndex={0}
      type="text"
      value=""
    />
    <div
      aria-live="polite"
      className="text"
      role="alert"
    >
      Select a category
    </div>
    <i
      aria-hidden="true"
      className="dropdown icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
      role="listbox"
    >
      <div
        aria-checked={false}
        aria-selected={true}
        className="selected item"
        onClick={[Function]}
        role="option"
        style={
          Object {
            "pointerEvents": "all",
          }
        }
      >
        <span
          className="text"
        >
          c1
        </span>
      </div>
      <div
        aria-checked={false}
        aria-selected={false}
        className="item"
        onClick={[Function]}
        role="option"
        style={
          Object {
            "pointerEvents": "all",
          }
        }
      >
        <span
          className="text"
        >
          c2
        </span>
      </div>
    </div>
  </div>
</span>
`;

exports[`<InDepthItemField /> Category Field 3`] = `
<div>
  category
  : 
  <div
    className="ui breadcrumb"
  >
    <div
      className="active section"
      onClick={[Function]}
    >
      test
    </div>
  </div>
</div>
`;

exports[`<InDepthItemField /> Description Field 1`] = `
<h1
  className="Title"
  style={
    Object {
      "marginLeft": "5%",
      "textAlign": "left",
    }
  }
>
  description
  :
   
  <div
    className="ui input"
  >
    <input
      name="description"
      onChange={[Function]}
      type="text"
      value={1}
    />
  </div>
</h1>
`;

exports[`<InDepthItemField /> Description Field 2`] = `
<h1
  className="Title"
  style={
    Object {
      "marginLeft": "5%",
      "textAlign": "left",
    }
  }
>
  1
</h1>
`;

exports[`<InDepthItemField /> Image Field 1`] = `
<div>
  <div
    style={
      Object {
        "height": 300,
        "width": 400,
      }
    }
  >
    <img
      className="ui image"
      onLoad={[Function]}
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAG8AAABYCAYAAAD7uql1AAAgAElEQVR4Xu29e5RV5Znu+5vfN6/r
VmtVFXWjoChABFERVCKgUUAjthFzMZqYxLjT3aP79D456d49zj5nZJzTo0f/0+fsHumT7O6xY/be
2umklcSEaFRuIiiIF0AUFQVFQSAlRRVVRdW6zut3/phrLs…"
    />
  </div>
  <button
    className="ui button"
    onClick={[Function]}
    role="button"
    style={
      Object {
        "display": "block",
        "margin": "0 auto",
      }
    }
  >
    <label
      className="Button"
      htmlFor="files"
    >
      Upload Image
    </label>
  </button>
  <div
    className="ui input"
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    <input
      id="files"
      onChange={[Function]}
      type="file"
      value=""
    />
  </div>
</div>
`;

exports[`<InDepthItemField /> Image Field 2`] = `
<div
  style={
    Object {
      "height": 300,
      "width": 400,
    }
  }
>
  <img
    className="ui image"
    onLoad={[Function]}
    src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAG8AAABYCAYAAAD7uql1AAAgAElEQVR4Xu29e5RV5Znu+5vfN6/r
VmtVFXWjoChABFERVCKgUUAjthFzMZqYxLjT3aP79D456d49zj5nZJzTo0f/0+fsHumT7O6xY/be
2umklcSEaFRuIiiIF0AUFQVFQSAlRRVVRdW6zut3/phrLs…"
  />
</div>
`;

exports[`<InDepthItemField /> Other Field 1`] = `
<span>
  other
  :
   
  <div
    className="ui input"
  >
    <input
      name="other"
      onChange={[Function]}
      type="text"
      value="other"
    />
  </div>
</span>
`;

exports[`<InDepthItemField /> Packaging Field 1`] = `
<span>
  packaging
  :
   
  <div
    aria-expanded={false}
    className="ui fluid search selection dropdown"
    onBlur={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseDown={[Function]}
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      autoComplete="off"
      className="search"
      onChange={[Function]}
      style={
        Object {
          "width": undefined,
        }
      }
      tabIndex={0}
      type="text"
      value=""
    />
    <div
      aria-live="polite"
      className="text"
      role="alert"
    />
    <i
      aria-hidden="true"
      className="dropdown icon"
      onClick={[Function]}
    />
    <div
      className="menu transition"
      role="listbox"
    >
      <div
        aria-checked={false}
        aria-selected={true}
        className="selected item"
        onClick={[Function]}
        role="option"
        style={
          Object {
            "pointerEvents": "all",
          }
        }
      >
        <span
          className="text"
        >
          p1
        </span>
      </div>
    </div>
  </div>
</span>
`;

exports[`<InDepthItemField /> Properties Field 1`] = `
<div>
  <span>
    Specifications
  </span>
  <ul>
    <li>
      test
       :
      <div
        className="ui input"
      >
        <input
          name="test"
          onChange={[Function]}
          type="text"
          value="1"
        />
      </div>
    </li>
  </ul>
</div>
`;

exports[`<InDepthItemField /> Properties Field 2`] = `
<div>
  <span>
    Specifications
  </span>
  <ul>
    <li>
      test
       : 
       
      1
       
      w
    </li>
  </ul>
</div>
`;

exports[`<InDepthItemField /> Properties Invalid Field 1`] = `
<div>
  <span>
    Specifications
  </span>
  <ul>
    <li>
      test
       :
      <div
        className="ui input"
      >
        <input
          name="test"
          onChange={[Function]}
          type="number"
        />
      </div>
    </li>
  </ul>
</div>
`;

exports[`<InDepthItemField /> Properties Invalid Field 2`] = `
<div>
  <span>
    Specifications
  </span>
  <ul>
    <li>
      test
       : 
      N/D
       
       
    </li>
  </ul>
</div>
`;

exports[`<InDepthItemField /> Renders properly 1`] = `"undefined: undefined"`;
